geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/ArrowBatchIterator.scala

```diff
@@ -0,0 +1,267 @@
+    * @param filter full filter for the query being run, used if querying enumeration values
+    * @param attributes names of attributes to dictionary encode
+    * @param provided provided dictionary values, if any, keyed by attribute name
+    * @return
+    */
+  def createDictionaries(ds: AccumuloDataStore,
+                         sft: SimpleFeatureType,
+                         filter: Option[Filter],
+                         attributes: Seq[String],
+                         provided: Map[String, Seq[AnyRef]]): Map[String, ArrowDictionary] = {
+    def name(i: Int): String = sft.getDescriptor(i).getLocalName
+
+    if (attributes.isEmpty) { Map.empty } else {
+      // note: sort values to return same dictionary cache
+      val providedDictionaries = provided.map { case (k, v) => k -> ArrowDictionary.create(sort(v)) }
+      val toLookup = attributes.filterNot(provided.contains)
+      val lookedUpDictionaries = if (toLookup.isEmpty) { Map.empty } else {
+        // use topk if available, otherwise run a live stats query to get the dictionary values
+        val read = ds.stats.getStats[TopK[AnyRef]](sft, toLookup).map { k =>
+          name(k.attribute) -> ArrowDictionary.create(sort(k.topK(1000).map(_._1)))
+        }.toMap
+        val toQuery = toLookup.filterNot(read.contains)
+        val queried = if (toQuery.isEmpty) { Map.empty } else {
+          val stats = Stat.SeqStat(toQuery.map(Stat.Enumeration))
+          val enumerations = ds.stats.runStats[EnumerationStat[String]](sft, stats, filter.getOrElse(Filter.INCLUDE))
+          enumerations.map { e => name(e.attribute) -> ArrowDictionary.create(sort(e.values.toSeq)) }.toMap
+        }
+        queried ++ read
+      }
+      providedDictionaries ++ lookedUpDictionaries
+    }
+  }
+
+  private def sort(values: Seq[AnyRef]): Seq[AnyRef] = {
+    if (values.isEmpty || !classOf[Comparable[_]].isAssignableFrom(values.head.getClass)) {
+      values
+    } else {
+      implicit val ordering = new Ordering[AnyRef] {
+        def compare(left: AnyRef, right: AnyRef): Int = left.asInstanceOf[Comparable[AnyRef]].compareTo(right)
+      }
+      values.sorted
+    }
+  }
+
+  /**
+    * Adapts the iterator to create simple features.
+    * WARNING - the same feature is re-used and mutated - the iterator stream should be operated on serially.
+    */
+  def kvsToFeatures(): (Entry[Key, Value]) => SimpleFeature = {
+    val sf = new ScalaSimpleFeature("", ArrowEncodedSft)
+    sf.setAttribute(1, GeometryUtils.zeroPoint)
+    (e: Entry[Key, Value]) => {
+      sf.setAttribute(0, e.getValue.get())
+      sf
+    }
+  }
+
+  /**
+    * First feature contains metadata for arrow file and dictionary batch, subsequent features
+    * contain record batches, final feature contains EOF indicator
+    *
+    * @param sft simple feature types
+    * @param hints query hints
+    * @param dictionaries dictionaries
+    * @return
+    */
+  def reduceFeatures(sft: SimpleFeatureType,
```

Comment:
there are 2 different options - this 'batch' iterator requires the dictionaries to be computed up front, which is done in the query planner. It returns a single logical arrow file, with a header, dictionaries, then batches of records. Each iterator instance creates a batch. There are a few options for computing dictionaries - you can define the values you want to encode, and anything else will be 'other', or you use stats topk (if available), or it will do a live stats query to determine the full dictionary.
The alternative method is the 'file' iterator, which will generate separate logical arrow from each iterator. In this case the dictionaries will be built up by each iterator, but they don't get merged. 
Our readers can handle both cases.